File: program/Hoi4 Parser/parse_tree.cpp
#include "parse_tree.h"
#include "token_types.h"
#include "use_ex_log.h"
#include <format>

using namespace std;
using namespace hoi4::parser;

const char ParseTree::openb = '{';
const char ParseTree::closb = '}';
const char ParseTree::equal = '=';
const char ParseTree::gter = '>';
const char ParseTree::less = '<';

// extern hoi4::ExLog ex_log()->append;
const char* fn_tree = "parse_tree";
#define UNKNOWN_ERROR ex_log()->append(fn_tree, format( "unknown error at line({}), column({})", _e.line(), _e.column()).c_str(), ExLog::ERR)
#define UNEXPECTED_KEY ex_log()->append(fn_tree, format( "unexpected key at line({}), column({})", _e.line(), _e.column()).c_str(), ExLog::ERR)
#define UNEXPECTED_OPERATOR ex_log()->append(fn_tree, format( "unexpected operator at line({}), column({})", _e.line(), _e.column()).c_str(), ExLog::ERR)
#define UNEXPECTED_VALUE ex_log()->append(fn_tree, format( "unexpected value at line({}), column({})", _e.line(), _e.column()).c_str(), ExLog::ERR)
#define UNEXPECTED_ARRAY_TYPE ex_log()->append(fn_tree, format( "unexpected array type at line({}), column({})", _e.line(), _e.column()).c_str(), ExLog::ERR)
#define ERROR_SYNTAX_ARRAY ex_log()->append(fn_tree, format( "wrong array syntax at line({}), column({})", _e.line(), _e.column()).c_str(), ExLog::ERR)


ParseTree::ParseTree() :
	key(nullptr),
	op(nullptr),
	value(nullptr),
	arr(nullptr),
	build(nullptr),
	from(nullptr),
	curr_sub(nullptr),
	step(NONE),
	level(0)
{
}

// call for sub-tree
ParseTree::ParseTree(const pTree _from, Value _key, Value _op, const size_t& _level) :
	key(_key),
	op(_op),
	value(nullptr),
	arr(nullptr),
	build(nullptr),
	from(_from),
	curr_sub(nullptr),
	step(OP),
	level(_level)
{
}

pToken ParseTree::once_get()
{
	return build.release();
}

const pTree ParseTree::get_from()
{
	return from;
}

void ParseTree::append(pToken _t)
{
	pScope(build.get())->append(_t);
}

const pTree ParseTree::parse(Element& _e)
{
	const char& ch = _e.head();
	if (step & SUB)
	{
		return par_sub(_e);
	}
	else if (step & ARR)
	{
		return par_arr(_e);
	}
	else if (step & OP)
	{
		//
		// 5
		//
		if (step & ON)
		{
			switch (ch)
			{
			case equal:
			case gter:
			case less:
				UNEXPECTED_VALUE;
				_e.value().reset();
				return from;
			case closb:
				_e.value().reset();
				done();
				return from;
			case openb:
				step = Steps(ARR);
				_e.value().reset();
				return pTree(this);
			default:
				step = Steps(SUB | KEY);
				value.reset(_e.value().release());
				build.reset(new Scope(key, level));
				return pTree(this);
			}
		}
		//
		// 2 - OP
		//
		else
		{
			switch (ch)
			{
			case closb:
			case equal:
			case gter:
			case less:
				UNEXPECTED_VALUE;
				_e.value().reset();
				return from;
			case openb:
				if (op[0] != equal)
				{
					UNEXPECTED_OPERATOR;
					_e.value().reset();
					return from;
				}
				else
				{
					step = Steps(OP | ON);
					_e.value().reset();
					return pTree(this);
				}
			default:
				step = Steps(VAL);
				value.reset(_e.value().release());
				build.reset(new Tag(key, op, value, level));
				return pTree(this);
			}
		}
	}
	//
	// 1
	//
	else if (step & KEY)
	{
		switch (ch)
		{
		case equal:
		case gter:
		case less:
			step = OP;
			op.reset(_e.value().release());
			return pTree(this);
		default:
			UNEXPECTED_OPERATOR;
			_e.value().reset();
			return from;
		}
	}
	//
	// 3
	//
	else if (step & VAL)
	{
		switch (ch)
		{
		case openb:
			step = TAG;
			_e.value().reset();
			return pTree(this);
		default:
			done();
			// _e.value().reset(); // leave _e to next tree
			return from;
		}
	}
	//
	// 4
	//
	else if (step & TAG)
	{
		switch (ch)
		{
		case openb:
		case equal:
		case gter:
		case less:
			UNEXPECTED_VALUE;
			_e.value().reset();
			return from;
		case closb:
			_e.value().reset();
			done();
			return from;
		default:
			step = TAG;
			pTag(build.get())->append(_e.value());
			return pTree(this);
		}
	}
	//
	// 0 - None
	//
	else
	{
		switch (ch)
		{
		case openb:
		case closb:
		case equal:
		case gter:
		case less:
			UNEXPECTED_KEY;
			_e.value().reset();
			return from;
		default:
			step = KEY;
			key.reset(_e.value().release());
			return pTree(this);
		}
	}
}

const pTree ParseTree::par_sub(Element& _e)
{
	const char& ch = _e.head();
	//
	// 6
	//
	if (step & KEY)
	{
		switch (ch)
		{
		case openb:
			UNEXPECTED_VALUE;
			_e.value().reset();
			return from;
		case closb:
			pScope(build.get())->append(new Token(value, level + 1));
			_e.value().reset();
			done();
			return from;
		case equal:
		case gter:
		case less:
			step = SUB;
			curr_sub = new ParseTree(pTree(this), value.release(), _e.value().release(), level + 1);
			return curr_sub;
		default:
			// step = Steps(SUB | KEY);
			pScope(build.get())->append(new Token(value, level + 1));
			value.reset(_e.value().release());
			return pTree(this);
		}
	}
	//
	// 7 - SUB
	//
	else
	{
		switch (ch)
		{
		case openb:
		case closb:
		case equal:
		case gter:
		case less:
			delete curr_sub;
			_e.value().reset();
			done();
			return from;
		default:
			delete curr_sub;
			step = Steps(SUB | KEY);
			value.reset(_e.value().release());
			return pTree(this);
		}
	}

}

const pTree ParseTree::par_arr(Element& _e)
{
	const char& ch = _e.head();
	if (step & TAG)
	{
		return par_tag_arr(_e);
	}
	else if (step & VAL)
	{
		// 11 - 14
		return par_val_arr(_e);
	}
	//
	// 9
	//
	if (step & OFF)
	{
		switch (ch)
		{
		case openb:
			step = ARR;
			_e.value().reset();
			return pTree(this);
		case closb:
			_e.value().reset();
			done();
			return from;
		default:
			ERROR_SYNTAX_ARRAY;
			_e.value().reset();
			return from;
		}
	}
	//
	// 10
	//
	else if (step & KEY)
	{
		switch (ch)
		{
		case openb:
			UNEXPECTED_VALUE;
			_e.value().reset();
			return from;
		case gter:
		case less:
			UNEXPECTED_OPERATOR;
			_e.value().reset();
			return from;
		case equal:
			step = Steps(ARR | TAG);
			build.reset(new TagArray(key, level));
			pTagArr(build.get())->append_new(arr);
			_e.value().reset();
			return pTree(this);
		case closb:
			step = Steps(ARR | VAL | OFF);
			build.reset(new ValueArray(key, level));
			pValArr(build.get())->append_new(arr);
			_e.value().reset();
			return pTree(this);
		default:
			step = Steps(ARR | VAL);
			build.reset(new ValueArray(key, level));
			pValArr(build.get())->append_new(arr);
			arr.reset(_e.value().release());
			pValArr(build.get())->append(arr);
			return pTree(this);
		}
	}
	//
	// 8 - ARR
	//
	else
	{
		switch (ch)
		{
		case openb:
		case equal:
		case gter:
		case less:
			UNEXPECTED_VALUE;
			_e.value().reset();
			return from;
		case closb:
			step = Steps(ARR | OFF);
			_e.value().reset();
			return pTree(this);
		default:
			step = Steps(ARR | KEY);
			arr.reset(_e.value().release());
			return pTree(this);
		}
	}
}

const pTree ParseTree::par_val_arr(Element& _e)
{
	const char& ch = _e.head();
	//
	// 12
	//
	if (step & OFF)
	{
		switch (ch)
		{
		case openb:
			step = Steps(ARR | VAL | ON);
			_e.value().reset();
			return pTree(this);
		case closb:
			_e.value().reset();
			done();
			return from;
		default:
			ERROR_SYNTAX_ARRAY;
			_e.value().reset();
			return from;
		}
	}
	//
	// 13
	//
	else if (step & ON)
	{
		switch (ch)
		{
		case openb:
		case equal:
		case gter:
		case less:
			UNEXPECTED_VALUE;
			_e.value().reset();
			return from;
		case closb:
			step = Steps(ARR | VAL | OFF);
			_e.value().reset();
			return pTree(this);
		default:
			step = Steps(ARR | VAL | KEY);
			arr.reset(_e.value().release());
			pValArr(build.get())->append_new(arr);
			return pTree(this);
		}
	}
	//
	// 14
	//
	else if (step & KEY)
	{
		switch (ch)
		{
		case openb:
		case equal:
		case gter:
		case less:
			UNEXPECTED_ARRAY_TYPE;
			_e.value().reset();
			return from;
		case closb:
			step = Steps(ARR | VAL | OFF);
			_e.value().reset();
			return pTree(this);
		default:
			step = Steps(ARR | VAL);
			arr.reset(_e.value().release());
			pValArr(build.get())->append(arr);
			return pTree(this);
		}
	}
	//
	// 11 - ARR | VAL
	//
	else
	{
		switch (ch)
		{
		case openb:
		case equal:
		case gter:
		case less:
			UNEXPECTED_VALUE;
			_e.value().reset();
			return from;
		case closb:
			step = Steps(ARR | VAL | OFF);
			_e.value().reset();
			return pTree(this);
		default:
			// step = Steps(ARR | VAL);
			arr.reset(_e.value().release());
			pValArr(build.get())->append(arr);
			return pTree(this);
		}
	}
}

const pTree ParseTree::par_tag_arr(Element& _e)
{

	const char& ch = _e.head();
	if (step & VAL)
	{
		//
		// 17
		//
		if (step & OFF)
		{
			switch (ch)
			{
			case openb:
			case equal:
			case gter:
			case less:
				UNEXPECTED_KEY;
				_e.value().reset();
				return from;
			case closb:
				step = Steps(ARR | TAG | OFF); //19
				_e.value().reset();
				return pTree(this);
			default:
				step = Steps(ARR | TAG | KEY);
				arr.reset(_e.value().release());
				pTagArr(build.get())->append_tag(arr);
				return pTree(this);
			}
		}
		//
		// 16 - ARR | TAG | VAL
		//
		else
		{
			switch (ch)
			{
			case openb:
			case equal:
			case gter:
			case less:
				UNEXPECTED_VALUE;
				_e.value().reset();
				return from;
			case closb:
				step = Steps(ARR | TAG | VAL | OFF);
				_e.value().reset();
				return pTree(this);
			default:
				// step = Steps(ARR | TAG | VAL);
				arr.reset(_e.value().release());
				pTagArr(build.get())->append(arr);
				return pTree(this);
			}
		}
	}
	//
	// 18
	//
	else if (step & OFF)
	{
		switch (ch)
		{
		case openb:
			step = Steps(ARR | TAG | ON);
			_e.value().reset();
			return pTree(this);
		case closb:
			_e.value().reset();
			done();
			return from;
		default:
			ERROR_SYNTAX_ARRAY;
			_e.value().reset();
			return from;
		}
	}
	//
	// 19
	//
	if (step & ON)
	{
		switch (ch)
		{
		case openb:
		case equal:
		case gter:
		case less:
			UNEXPECTED_KEY;
			_e.value().reset();
			return from;
		case closb:
			step = Steps(ARR | TAG | OFF);
			_e.value().reset();
			return pTree(this);
		default:
			step = Steps(ARR | TAG | KEY);
			arr.reset(_e.value().release());
			pTagArr(build.get())->append_new(arr);
			return pTree(this);
		}
	}
	//
	// 20
	//
	else if (step & KEY)
	{
		switch (ch)
		{
		case equal:
			step = Steps(ARR | TAG);
			_e.value().reset();
			return pTree(this);
		case gter:
		case less:
			UNEXPECTED_OPERATOR;
			_e.value().reset();
			return from;
		default:
			UNEXPECTED_ARRAY_TYPE;
			_e.value().reset();
			return from;
		}
	}
	//
	// 15 - ARR | TAG
	//
	else
	{
		switch (ch)
		{
		case openb:
			step = Steps(ARR | TAG | VAL);
			_e.value().reset();
			return pTree(this);
		default:
			UNEXPECTED_ARRAY_TYPE;
			_e.value().reset();
			return from;
		}
	}
}

void ParseTree::done()
{
	if (from != nullptr)
	{
		from->append(build.release());
	}
}

File: program/Parser/ParseTree.cs
using System;
using System.Collections.Generic;
using System.Linq;
using System.Text;
using System.Threading.Tasks;

namespace Parser
{
    public class ParseTree
    {
        const char OpenBrace = '{';
        const char CloseBrace = '}';
        const char Equal = '=';
        const char Greater = '>';
        const char Less = '<';

        enum Steps
        {
            None = 0,
            Key = 0b1,
            Operator = 0b1 << 1,
            Value = 0b1 << 2,
            Tag = 0b1 << 3,
            Array = 0b1 << 4,
            Sub = 0b1 << 5,
            On = 0b1 << 6,
            Off = 0b1 << 7
        }
        Steps Step { get; set; } = Steps.None;
        string Key { get; set; } = string.Empty;
        string Operator { get; set; } = string.Empty;
        string Value { get; set; } = string.Empty;
        string Array { get; set; } = string.Empty;
        TokenAPI Builder { get; set; } = null;
        public ParseTree From { get; init; }
        uint Level { get; init; }
        bool OwnBuilder { get; set; } = true;

        public ParseTree()
        {
            From = null;
            Level = 0;
        }

        public ParseTree(ParseTree from, uint level, string key, string operat)
        {
            From = from;
            Level = level;
            Key = key;
            Operator = operat;
        }

        public TokenAPI OnceGet()
        {
            if (OwnBuilder)
            {
                OwnBuilder = false;
                return Builder;
            }
            else { return null; }
        }
        private void Done()
        {
            From?.Append(Builder);
        }

        public void Append(TokenAPI token)
        {
            ((Scope)Builder).Append(token);
        }

        public ParseTree Parse(Element element)
        {
            char ch = element.Head();
            if ((Step & Steps.Sub) > 0)
            {
                return Parse_Sub(element);
            }
            else if ((Step & Steps.Array) > 0)
            {
                return Parse_Array(element);
            }
            else if ((Step & Steps.Operator) > 0)
            {
                return Parse_Operator(element);
            }
            //
            // 1
            //
            else if ((Step & Steps.Key) > 0)
            {
                switch(ch)
                {
                    case Equal:
                    case Greater:
                    case Less:
                        Step = Steps.Operator;
                        Operator = element.GetValue();
                        return this;
                    default:
                        Exceptions.UnexpectedOperator(element);
                        element.GetValue();
                        return From;
                }
            }
            //
            // 3
            //
            else if ((Step & Steps.Value) > 0)
            {
                switch(ch)
                {
                    case OpenBrace:
                        Step = Steps.Value;
                        element.GetValue();
                        return this;
                    default:
                        Done();
                        // element.GetValue(); // leave element to next tree
                        return From;
                }
            }
            //
            // 4
            //
            else if ((Step & Steps.Tag) > 0)
            {
                switch(ch)
                {
                    case OpenBrace:
                    case Equal:
                    case Greater:
                    case Less:
                        Exceptions.UnexpectedValue(element);
                        element.GetValue();
                        return From;
                    case CloseBrace:
                        element.GetValue();
                        Done();
                        return From;
                    default:
                        Step = Steps.Tag;
                        ((TaggedValue)Builder).Append(Value);
                        return this;
                }
            }
            //
            // 0 - None
            //
            else
            {
                switch(ch)
                {
                    case OpenBrace:
                    case CloseBrace:
                    case Equal:
                    case Greater:
                    case Less:
                        Exceptions.UnexpectedKey(element);
                        element.GetValue();
                        return From;
                    default:
                        Step = Steps.Key;
                        Key = element.GetValue();
                        return this;
                }
            }
        }
        public ParseTree Parse_Sub(Element element)
        {
            char ch = element.Head();
            //
            // 6
            //
            if ((Step & Steps.Key) > 0)
            {
                switch (ch)
                {
                    case OpenBrace:
                        Exceptions.UnexpectedValue(element);
                        element.GetValue();
                        return From;
                    case CloseBrace:
                        ((Scope)Builder).Append(new Token(Key, Level + 1));
                        element.GetValue();
                        Done();
                        return From;
                    case Equal:
                    case Greater:
                    case Less:
                        Step = Steps.Sub;
                        return new ParseTree(this, Level + 1, Key, element.GetValue());
                    default:
                        ((Scope)Builder).Append(new Token(Value, Level + 1));
                        Value = element.GetValue();
                        return this;
                }
            }
            //
            // 7 - Sub
            //
            else
            {
                switch (ch)
                {
                    case OpenBrace:
                    case CloseBrace:
                    case Equal:
                    case Greater:
                    case Less:
                        element.GetValue();
                        Done();
                        return From;
                    default:
                        Step = Steps.Sub | Steps.Key;
                        element.GetValue();
                        return this;
                }
            }
        }
        public ParseTree Parse_Array(Element element)
        {
            char ch = element.Head();
            if ((Step & Steps.Tag) > 0)
            { 
                return Parse_Tag_Array(element);
            }
            else if ((Step & Steps.Value) > 0)
            {
                return Parse_Value_Array(element);
            }
            //
            // 9
            //
            else if ((Step & Steps.Off) > 0)
            {
                switch(ch)
                {
                    case OpenBrace:
                        Step = Steps.Array;
                        element.GetValue();
                        return this;
                    case CloseBrace:
                        element.GetValue();
                        Done();
                        return From;
                    default:
                        Exceptions.UnexpectedArraySyntax(element);
                        element.GetValue();
                        return From;
                }
            }
            //
            // 10
            //
            else if ((Step & Steps.Key) > 0)
            {
                switch (ch)
                {
                    case OpenBrace:
                        Exceptions.UnexpectedValue(element);
                        element.GetValue();
                        return From;
                    case Greater:
                    case Less:
                        Exceptions.UnexpectedOperator(element);
                        element.GetValue();
                        return From;
                    case Equal:
                        Step = Steps.Array | Steps.Tag;
                        Builder = new TagArray(Key, Level);
                        ((TagArray)Builder).Append_new(Array);
                        element.GetValue();
                        return this;
                    case CloseBrace:
                        Step = Steps.Array | Steps.Value | Steps.Off;
                        Builder = new ValueArray(Key, Level);
                        ((ValueArray)Builder).Append_new(Array);
                        element.GetValue();
                        return this;
                    default:
                        Step = Steps.Array | Steps.Value;
                        Builder = new ValueArray(Key, Level);
                        ((ValueArray)Builder).Append_new(Array);
                        ((ValueArray)Builder).Append(element.GetValue());
                        return this;
                }
            }
            //
            // 8 - Array
            //
            else
            {
                switch(ch)
                {
                    case OpenBrace:
                    case Equal:
                    case Greater:
                    case Less:
                        Exceptions.UnexpectedValue(element);
                        element.GetValue();
                        return From;
                    case CloseBrace:
                        Step = Steps.Array | Steps.Off;
                        element.GetValue();
                        return this;
                    default:
                        Step = Steps.Array | Steps.Key;
                        Array = element.GetValue();
                        return this;
                }
            }
        }
        public ParseTree Parse_Tag_Array(Element element)
        {
            char ch = element.Head();
            //
            // 12
            //
            if ((Step & Steps.Off) > 0)
            {
                switch (ch)
                {
                    case OpenBrace:
                        Step = Steps.Array | Steps.Value | Steps.On;
                        element.GetValue();
                        return this;
                    case CloseBrace:
                        element.GetValue();
                        Done();
                        return From;
                    default:
                        Exceptions.UnexpectedArraySyntax(element);
                        element.GetValue();
                        return From;
                }
            }
            //
            // 13
            //
            else if ((Step & Steps.On) > 0)
            {
                switch(ch)
                {
                    case OpenBrace:
                    case Equal:
                    case Greater:
                    case Less:
                        Exceptions.UnexpectedValue(element);
                        element.GetValue(); 
                        return From;
                    case CloseBrace:
                        Step = Steps.Array | Steps.Value | Steps.Off;
                        element.GetValue();
                        return this;
                    default:
                        Step = Steps.Array | Steps.Value | Steps.Key;
                        ((ValueArray)Builder).Append_new(element.GetValue());
                        return this;
                }
            }
            //
            // 14
            //
            else if ((Step & Steps.Key) > 0)
            {
                switch (ch)
                {
                    case OpenBrace:
                    case Equal:
                    case Greater:
                    case Less:
                        Exceptions.UnexpectedArrayType(element);
                        element.GetValue(); 
                        return From;
                    case CloseBrace:
                        Step = Steps.Array | Steps.Value | Steps.Off;
                        element.GetValue();
                        return this;
                    default:
                        Step = Steps.Array | Steps.Value;
                        ((ValueArray)Builder).Append(element.GetValue());
                        return this;
                }
            }
            //
            // 11 - Array | Value
            //
            else
            {
                switch(ch)
                {
                    case OpenBrace:
                    case Equal:
                    case Greater:
                    case Less:
                        Exceptions.UnexpectedValue(element);
                        element.GetValue();
                        return From;
                    case CloseBrace:
                        Step = Steps.Array | Steps.Value | Steps.Off;
                        element.GetValue();
                        return this;
                    default:
                        ((ValueArray)Builder).Append(element.GetValue());
                        return this;
                }
            }
        }
        public ParseTree Parse_Value_Array(Element element)
        {
            char ch = element.Head();
            if ((Step & Steps.Value) > 0)
            {
                //
                // 17
                //
                if ((Step & Steps.Off) > 0)
                {
                    switch (ch)
                    {
                        case OpenBrace:
                        case Equal:
                        case Greater:
                        case Less:
                            Exceptions.UnexpectedKey(element);
                            element.GetValue();
                            return From;
                        case CloseBrace:
                            Step = Steps.Array | Steps.Tag | Steps.Off;
                            element.GetValue();
                            return this;
                        default:
                            Step = Steps.Array | Steps.Tag | Steps.Key;
                            ((TagArray)Builder).Append(element.GetValue());
                            return this;
                    }
                }
                //
                // 16 - Array | Tag | Value
                //
                else
                {
                    switch (ch)
                    {
                        case OpenBrace:
                        case Equal:
                        case Greater:
                        case Less:
                            Exceptions.UnexpectedValue(element);
                            element.GetValue();
                            return From;
                        case CloseBrace:
                            Step = Steps.Array | Steps.Tag | Steps.Value | Steps.Off;
                            element.GetValue();
                            return this;
                        default:
                            element.GetValue();
                            ((TagArray)Builder).Append(element.GetValue());
                            return this;
                    }
                }
            }
            //
            // 18
            //
            else if ((Step & Steps.Off) > 0)
            {
                switch(ch)
                {
                    case OpenBrace:
                        Step = Steps.Array | Steps.Tag | Steps.On;
                        element.GetValue();
                        return this;
                    case CloseBrace:
                        element.GetValue();
                        Done();
                        return From;
                    default:
                        Exceptions.UnexpectedArraySyntax(element);
                        element.GetValue();
                        return From;
                }
            }
            //
            // 19
            //
            else if ((Step & Steps.On) > 0)
            {
                switch(ch)
                {
                    case OpenBrace:
                    case Equal:
                    case Greater:
                    case Less:
                        Exceptions.UnexpectedKey(element);
                        element.GetValue();
                        return From;
                    case CloseBrace:
                        Step = Steps.Array | Steps.Tag | Steps.Off;
                        element.GetValue();
                        return this;
                    default:
                        Step = Steps.Array | Steps.Tag | Steps.Key;
                        ((TagArray)Builder).Append_new(element.GetValue());
                        return this;
                }
            }
            //
            // 20
            //
            else if ((Step & Steps.Key) > 0)
            {
                switch(ch)
                {
                    case Equal:
                        Step = Steps.Array | Steps.Tag;
                        element.GetValue();
                        return this;
                    case Greater:
                    case Less:
                        Exceptions.UnexpectedOperator(element);
                        element.GetValue();
                        return From;
                    default:
                        Exceptions.UnexpectedArrayType(element);
                        element.GetValue();
                        return From;
                }
            }
            //
            //
            // 15 - Array | Tag
            else
            {
                switch(ch)
                {
                    case OpenBrace:
                        Step = Steps.Array | Steps.Tag | Steps.Value;
                        element.GetValue();
                        return this;
                    default:
                        Exceptions.UnexpectedArrayType(element);
                        element.GetValue();
                        return From;
                }
            }
        }

        #region ==== Parse Operator (5, 2) ====

        public ParseTree Parse_Operator(Element element)
        {
            char ch = element.Head();
            //
            // 5
            //
            if ((Step & Steps.On) > 0)
            {
                switch (ch)
                {
                    case Equal:
                    case Greater:
                    case Less:
                        Exceptions.UnexpectedValue(element);
                        element.GetValue();
                        return From;
                    case CloseBrace:
                        element.GetValue();
                        Done();
                        return From;
                    case OpenBrace:
                        Step = Steps.Array;
                        element.GetValue();
                        return this;
                    default:
                        Step = Steps.Sub | Steps.Key;
                        Value = element.GetValue();
                        Builder = new Scope(Key, Level);
                        return this;
                }
            }
            //
            // 2 - Op
            //
            else
            {
                switch (ch)
                {
                    case CloseBrace:
                    case Equal:
                    case Greater:
                    case Less:
                        Exceptions.UnexpectedValue(element);
                        element.GetValue();
                        return From;
                    case OpenBrace:
                        if (Operator[0] != Equal)
                        {
                            Exceptions.UnexpectedOperator(element);
                            element.GetValue();
                            return From;
                        }
                        else
                        {
                            Step = Steps.Operator | Steps.On;
                            element.GetValue();
                            return this;
                        }
                    default:
                        Step = Steps.Value;
                        Builder = new TaggedValue(Key, Level, Operator, element.GetValue());
                        return this;
                }
            }
        }

        #endregion
    }
}
